File: src/types/workspace.ts
import type { Project } from "@/types/project";
import type { ProjectScheduledTask } from "@/types/project-scheduled-task";
import type { ProjectWorker } from "@/types/project-worker";
import type { ServiceName, ServiceState } from "@/types/service";

export interface BootState {
  appName: string;
  environment: "tauri" | "browser";
  dbPath: string;
  startedAt: string;
}

export interface WorkspacePortStatus {
  port: number;
  available: boolean;
  pid?: number | null;
  processName?: string | null;
  managedOwner?: ServiceName | null;
  expectedServices: ServiceName[];
}

export interface WorkspaceOverviewPayload {
  bootState: BootState;
  projects: Project[];
  services: ServiceState[];
  workers: ProjectWorker[];
  scheduledTasks: ProjectScheduledTask[];
  portSummary: WorkspacePortStatus[];
}

File: src-tauri/src/models/workspace.rs
use crate::models::project::Project;
use crate::models::scheduled_task::ProjectScheduledTask;
use crate::models::service::{ServiceName, ServiceState};
use crate::models::worker::ProjectWorker;
use crate::state::BootState;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspacePortStatus {
    pub port: u16,
    pub available: bool,
    pub pid: Option<u32>,
    pub process_name: Option<String>,
    pub managed_owner: Option<ServiceName>,
    pub expected_services: Vec<ServiceName>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceOverviewPayload {
    pub boot_state: BootState,
    pub projects: Vec<Project>,
    pub services: Vec<ServiceState>,
    pub workers: Vec<ProjectWorker>,
    pub scheduled_tasks: Vec<ProjectScheduledTask>,
    pub port_summary: Vec<WorkspacePortStatus>,
}
